File: src/main/python/candy/Candy.py
class Solution:
    def candy(self, ratings: list[int]) -> int:
        n = len(ratings)
        left2right = [1] * n
        right2left = [1] * n
        for i in range(1, n):
            if ratings[i - 1] >= ratings[i]:
                left2right[i] = 1
            else:
                left2right[i] = left2right[i - 1] + 1
        for i in range(n - 2, -1, -1):
            if ratings[i] <= ratings[i + 1]:
                right2left[i] = 1
            else:
                right2left[i] = right2left[i + 1] + 1
        return sum([max(a, b) for a, b in zip(left2right, right2left)])

    def test(self):
        tests = [
            [[1, 0, 2], 5],
            [[1, 2, 2], 4],
            [[1, 3, 2, 2, 1], 7],
            [[1, 2, 87, 87, 87, 2, 1], 13]
        ]
        for test in tests:
            ratings, expect_result = test
            result = self.candy(ratings)
            print(result)
            print(result == expect_result)


if __name__ == '__main__':
    Solution().test()

File: src/main/kotlin/me/bossm0n5t3r/leetcode/candy/Candy.kt
package me.bossm0n5t3r.leetcode.candy

class Candy {
    class Solution {
        fun candy(ratings: IntArray): Int {
            val left2Right = IntArray(ratings.size) { 1 }
            val right2Left = IntArray(ratings.size) { 1 }
            for (i in 1 until ratings.size) {
                left2Right[i] = if (ratings[i - 1] >= ratings[i]) 1 else left2Right[i - 1] + 1
            }
            for (i in ratings.size - 2 downTo 0) {
                right2Left[i] = if (ratings[i + 1] >= ratings[i]) 1 else right2Left[i + 1] + 1
            }
            return left2Right.zip(right2Left).sumOf { maxOf(it.first, it.second) }
        }
    }
}
